[LV] Vectorize histogram operations

huntergr-arm opened the pull request:
This patch implements autovectorization support for the 'all-in-one' histogram intrinsic, which seems to have more support than the 'standalone' intrinsic. See https://discourse.llvm.org/t/rfc-vectorization-support-for-histogram-count-operations/74788/ for an overview of the work and my notes on the tradeoffs between the two approaches.

This is a new PR after the previous one (#91458) was reverted at maintainer request.

I have removed the changes to LoopAccessAnalysis in favor continuing analysis in LoopVectorizationLegality.

huntergr-arm committed:
Initial tests for histogram autovec

huntergr-arm committed:
LV Changes

huntergr-arm committed:
Extra tests

paschalis-mpeis reviewed (commented):
I was only able to take a quick look at the changes; they do look good.

SamTebbs33 reviewed (approved):
Looks good to me as well, thank you.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -4622,6 +4622,10 @@ bool LoopVectorizationPlanner::isCandidateForEpilogueVectorization(
   if (OrigLoop->getExitingBlock() != OrigLoop->getLoopLatch())
     return false;
 
+  // Loops containing histograms are not currently supported.
+  if (!Legal->getHistograms().empty())
+    return false;
+
```
```suggestion
return Legal->getHistograms().empty()
```

huntergr-arm replied:
done

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8173,6 +8202,36 @@ VPWidenRecipe *VPRecipeBuilder::tryToWiden(Instruction *I,
+  if (Legal->isMaskRequired(HI->Store))
+    Mask = getBlockInMask(HI->Store->getParent());
+  else
+    Mask = Plan.getOrAddLiveIn(
```
For loads/stores the mask is optional, instead of providing an all true mask can we instead not add a mask if there's no mask needed?

huntergr-arm replied:
done. The intrinsic requires a mask, but the recipe doesn't. The all-true mask is now synthesized in the execute method.

I also improved recipe printing to use 'dec' instead of 'inc' if the opcode was a subtract.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8563,6 +8626,15 @@ LoopVectorizationPlanner::tryToBuildVPlanWithVPRecipes(VFRange &Range) {
+      // Note that if the initial VF is scalar, we need to generate the normal
+      // clone recipe for these instructions. A histogram recipe will only be
+      // generated when minVF > 1.
+      if (Legal->getHistogramInfo(Instr) && !isa<StoreInst>(Instr) &&
```
Would it be sufficient to leave to load around, to be cleaned up later by DCE?

huntergr-arm replied:
done

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -1573,6 +1575,39 @@ class VPWidenCallRecipe : public VPSingleDefRecipe {
 #endif
 };
 
+class VPHistogramRecipe : public VPRecipeBase {
```
please document the new recipe

huntergr-arm replied:
done

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -1573,6 +1575,39 @@ class VPWidenCallRecipe : public VPSingleDefRecipe {
 };
 
+class VPHistogramRecipe : public VPRecipeBase {
+  const HistogramInfo *Info;
```
can store by reference?

huntergr-arm replied:
done

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1054,6 +1058,110 @@ bool LoopVectorizationLegality::canVectorizeInstrs() {
+/// regardless of hardware support. When there is support, it additionally
+/// stores the BinOp/Load pairs in \p HistogramCounts, as well the pointers
+/// used to update histogram in \p HistogramPtrs.
+
```
nit: remove newline?

huntergr-arm replied:
done

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1054,6 +1058,110 @@ bool LoopVectorizationLegality::canVectorizeInstrs() {
+/// stores the BinOp/Load pairs in \p HistogramCounts, as well the pointers
+/// used to update histogram in \p HistogramPtrs.
+
+static bool findHistograms(LoadInst *LI, StoreInst *HSt, Loop *TheLoop,
```
nit: findHistograms as this only looks for a single one?

huntergr-arm replied:
done

fhahn commented on llvm/test/Transforms/LoopVectorize/histograms.ll:
```diff
@@ -0,0 +1,44 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 3
+; RUN: opt < %s -passes=loop-vectorize,instcombine -enable-histogram-loop-vectorization -S | FileCheck %s
```
would this vectorize with `-force-vector-width=2` or something like that?

huntergr-arm replied:
I added the flag, but it won't vectorize at present due to the cost being Invalid. Forcing the instruction cost to 1 no longer works, so this will have to wait for an improved cost model.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -958,6 +964,44 @@ void VPWidenCallRecipe::print(raw_ostream &O, const Twine &Indent,
+
+    // If this is a subtract, we want to invert the increment amount. We may
+    // add a separate intrinsic in future, but for now we'll try this.
+    if (Opcode == Instruction::Sub)
```
assert that opcode is :Add in the else?

huntergr-arm replied:
done

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1054,6 +1058,110 @@ bool LoopVectorizationLegality::canVectorizeInstrs() {
+  LoadInst *LI = dyn_cast<LoadInst>(Dep.getSource(DepChecker));
+  StoreInst *SI = dyn_cast<StoreInst>(Dep.getDestination(DepChecker));
+
+  if (!LI || !SI)
```
Just checking if we have a test where the source is a store and dest is a load, possibly for loops stepping backwards?

huntergr-arm replied:
We don't.

I tried creating one but was never able to get LAA to recognize the store as a source -- it always gave up before that point, possibly due to the indirection. Maybe I'm missing something.

huntergr-arm committed:
Addressing comments
* Mask is optional for vplan recipe
* Left load+update to DCE
* Documentation for new recipe
* Stored histogram info by reference
* Allow GEPs with more than one index value, provided that all
  but the last are constant.
* Cleanup to remove some unnecessary local variables.
* Minor nits addressed.

huntergr-arm committed:
Deal with change in dep type to IndirectUnsafe

huntergr-arm committed:
Add tailfolding histogram test case

huntergr-arm commented:
Rebased, addressed comments, added some more tests.

huntergr-arm commented:
@fhahn ping -- do the latest changes sufficiently address your concerns?

fhahn commented on llvm/include/llvm/Transforms/Vectorize/LoopVectorizationLegality.h:
```diff
@@ -390,6 +402,22 @@ class LoopVectorizationLegality {
+  }
+
+  /// Returns a list of all known histogram operations in the loop.
+  const SmallVectorImpl<HistogramInfo> &getHistograms() const {
```
looking at the users of this function, might be cleared to have `hasHistograms()` instead?

huntergr-arm replied:
done

fhahn commented on llvm/include/llvm/Transforms/Vectorize/LoopVectorizationLegality.h:
```diff
@@ -438,6 +466,11 @@ class LoopVectorizationLegality {
   bool canVectorizeMemory();
 
+  /// If LAA cannot determine whether all dependences are safe, we may be able
+  /// to further analyse some unknown dependences and if they match a certain
```
There was a recent LAA fix, and the dependences should now be classified as `IndirectUnsafe` I think. Would be good to update the comment

huntergr-arm replied:
done.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1046,6 +1050,114 @@ bool LoopVectorizationLegality::canVectorizeInstrs() {
+}
+
+bool LoopVectorizationLegality::canVectorizeUncheckedDependences() {
+  // For now, we only support an unknown dependency that calculates a histogram
```
unknown-> IndirectUnsafe?

huntergr-arm replied:
done.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1046,6 +1050,114 @@ bool LoopVectorizationLegality::canVectorizeInstrs() {
+  if (!EnableHistogramVectorization)
+    return false;
+
+  // FIXME: Support more than one unchecked dependence, and check to see if some
```
I think with the change to use IndirectUnsafe, unknown dependences should be fine, as we can generate runtime checks for those

huntergr-arm replied:
I hadn't appreciated that part of the LAA change. done, and I've added a test with a second dependency.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1046,6 +1050,114 @@ bool LoopVectorizationLegality::canVectorizeInstrs() {
+
+  const MemoryDepChecker::Dependence &Dep = (*Deps).front();
+
+  // We're only interested in Unknown or IndirectUnsafe dependences.
```
Is it sufficient to look for IndirectUnsafe dependences here?

huntergr-arm replied:
done.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -1575,6 +1577,52 @@ class VPWidenCallRecipe : public VPSingleDefRecipe {
+  /// Return the mask operand if one was provided, or a null pointer if all
+  /// lanes should be executed unconditionally.
+  VPValue *getMask() const {
+    if (getNumOperands() == 3)
+      return getOperand(2);
+    return nullptr;
```
```suggestion
   return getNumOperands() == 3 ? getOperand(2) : nullptr;
```

huntergr-arm replied:
done.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -958,6 +964,65 @@ void VPWidenCallRecipe::print(raw_ostream &O, const Twine &Indent,
+      Mask = Builder.CreateVectorSplat(
+          VTy->getElementCount(), ConstantInt::getTrue(Builder.getInt1Ty()));
+
+    // Not sure how to make IncAmt stay scalar yet. For now just extract the
```
Sufficient to implement `VPHistogramRecipe::onlyFirstLaneUsed` and return true for the inc operand?

huntergr-arm replied:
That didn't work, but supplying `true` to the IsScalar parameter of State.get did.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -958,6 +964,65 @@ void VPWidenCallRecipe::print(raw_ostream &O, const Twine &Indent,
+    if (Opcode == Instruction::Sub)
+      IncAmt = Builder.CreateNeg(IncAmt);
+    else
+      assert(Opcode == Instruction::Add);
```
```suggestion
      assert(Opcode == Instruction::Add && "only add or sub supported for now");
```

huntergr-arm replied:
done

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -30,6 +31,7 @@
 #include "llvm/Transforms/Utils/BasicBlockUtils.h"
 #include "llvm/Transforms/Utils/LoopUtils.h"
 #include "llvm/Transforms/Utils/ScalarEvolutionExpander.h"
+#include "llvm/Transforms/Vectorize/LoopVectorizationLegality.h"
```
Might be better to define HistogramInfo in VPlan.h instead of pulling `LoopVectorizationLegality.h` in here, as VPlanRecipes.cpp should not depend/rely on LVL

huntergr-arm replied:
Sadly, that's not quite possible at the moment. LoopVectorizationLegality.h needs the definition of HistogramInfo since it creates a SmallVector of them, and accesses the fields. I can't include VPlan.h in the other header, since VPlan.h is under the lib/ subtree instead of the include/ subtree. I could perhaps try to fudge it by always including VPlan.h first in any .cpp file, but that feels fragile (and clang-format doesn't like it anyway).

I suspect a shared header under include/ would be best. Or have I missed something?

fhahn replied:
It looks like VPHistogramRecipe only needs `Update` of `HistogramInfo` and that only for retrieving the legacy cost. Ideally we would include `VPHistogramRecipe::computeCost` in this patch, avoiding the need for accessing `Update`. 

If that isn't easily possibly, can we just store `Update` (with a comment what it is used for)?

fhahn replied:
Still pending?

huntergr-arm replied:
This was done with the last round of updates. Old diff?

fhahn reviewed (commented):
Thanks for the update! some additional comments inline

huntergr-arm committed:
Address comments, add tests for additional dependences.

huntergr-arm committed:
Rebase, require asserts for tests using debug output

huntergr-arm commented:
Rebased and added REQUIRES lines to tests using debug capabilities.

Any additional comments?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -274,6 +276,10 @@ static Instruction *getInstructionForCost(const VPRecipeBase *R) {
     return IG->getInsertPos();
   if (auto *WidenMem = dyn_cast<VPWidenMemoryRecipe>(R))
     return &WidenMem->getIngredient();
+  // FIXME: Override the cost method properly to take gather/scatter cost
```
Histogram cost computation looks self contained, can we instead directly implement `VPHistogramRecipe::computeCost` instead falling back?

huntergr-arm replied:
done

fhahn replied:
Thanks, can the FIXME & histogram case below be removed?

huntergr-arm replied:
They were removed. Are you looking at a cached version of the diff?

fhahn commented on llvm/test/Transforms/LoopVectorize/histograms.ll:
```diff
@@ -0,0 +1,44 @@
+
+target datalayout = "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-s0:64:64-f80:128:128-n8:16:32:64-S128"
+
+define void @simple_histogram(ptr noalias %buckets, ptr readonly %indices, i64 %N) #0 {
```
might be worth adding a comment here saying what we expect here, i.e. no vectorization as (generic) target doesn't support histogram intrinsics?

huntergr-arm replied:
done

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/sve2-histcnt.ll:
```diff
@@ -0,0 +1,973 @@
+; CHECK-NEXT:     WIDEN ir<%0> = load vp<%4>
+; CHECK-NEXT:     WIDEN-CAST ir<%idxprom1> = zext  ir<%0> to i64
+; CHECK-NEXT:     WIDEN-GEP Inv[Var] ir<%arrayidx2> = getelementptr inbounds ir<%buckets>, ir<%idxprom1>
+; CHECK-NEXT:     WIDEN-HISTOGRAM buckets: ir<%arrayidx2>, inc: ir<1>
```
It would probably better to have a separate test file to check printing of a plan with a histogram recipe. 

It should be sufficient to check printing for a single loop; Ideally also avoiding hard-coded `vp<%xx>` as this makes the tests fragile.

huntergr-arm replied:
done, `sve2-histcnt-vplan.ll`

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/sve2-histcnt.ll:
```diff
@@ -0,0 +1,973 @@
+  %arrayidx = getelementptr inbounds i32, ptr %indices, i64 %iv
+  %0 = load i32, ptr %arrayidx, align 4
+  %idxprom1 = zext i32 %0 to i64
+  %arrayidx2 = getelementptr inbounds i32, ptr %buckets, i64 %idxprom1
```
nit to possibly help readability 

 ```suggestion
  %gep.bucket  = getelementptr inbounds i32, ptr %buckets, i64 %idxprom1
```

huntergr-arm replied:
done

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/sve2-histcnt.ll:
```diff
@@ -0,0 +1,973 @@
+for.body:
+  %iv = phi i64 [ 0, %entry ], [ %iv.next, %for.body ]
+  %arrayidx = getelementptr inbounds i32, ptr %indices, i64 %iv
+  %0 = load i32, ptr %arrayidx, align 4
```
nit to possibly help readability



```suggestion
  %l.idx = load i32, ptr %arrayidx, align 4
```

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/sve2-histcnt.ll:
```diff
@@ -0,0 +1,973 @@
+
+for.body:
+  %iv = phi i64 [ 0, %entry ], [ %iv.next, %for.body ]
+  %arrayidx = getelementptr inbounds i32, ptr %indices, i64 %iv
```
nit to possibly help readability



```suggestion
  %gep.indices = getelementptr inbounds i32, ptr %indices, i64 %iv
```

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/sve2-histcnt.ll:
```diff
@@ -0,0 +1,973 @@
+  %0 = load i32, ptr %arrayidx, align 4
+  %idxprom1 = zext i32 %0 to i64
+  %arrayidx2 = getelementptr inbounds i32, ptr %buckets, i64 %idxprom1
+  %1 = load i32, ptr %arrayidx2, align 4
```
nit to possibly help readability
```suggestion
  %l.bucket = load i32, ptr %arrayidx2, align 4
```

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1048,6 +1052,123 @@ bool LoopVectorizationLegality::canVectorizeInstrs() {
+  LAI = &LAIs.getInfo(*TheLoop);
+  const MemoryDepChecker::Dependence *IUDep = nullptr;
+  const MemoryDepChecker &DepChecker = LAI->getDepChecker();
+  const auto *Deps = DepChecker.getDependences();
```
`Deps` may be nullptr here; LAA only records dependences up to `MaxDependences` (would be good to have a test using something like `-max-dependences=2`

huntergr-arm replied:
Done, `sve2-histcnt-too-many-deps.ll`

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1048,6 +1052,123 @@ bool LoopVectorizationLegality::canVectorizeInstrs() {
+  // any extensions.
+  // FIXME: Support indices from other sources that a linear load from memory?
+  Value *VPtrVal;
+  if (!match(HIdx, m_ZExtOrSExtOrSelf(m_Load(m_Value(VPtrVal)))))
```
Do we have a tests where the index isn't extended? Coudlnt' find any in  sve2-histcnt.ll but might have missed them

huntergr-arm replied:
I've added `simple_histogram_64b`.

fhahn replied:
Thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1048,6 +1052,123 @@ bool LoopVectorizationLegality::canVectorizeInstrs() {
+
+  // Restrict address calculation to constant indices except for the last term.
+  Value *HIdx = nullptr;
+  for (Value *Index : GEP->indices()) {
```
Are all cases exercised in tests? Looks like the tests are all using `%arrayidx2 = getelementptr inbounds i32, ptr %buckets, i64 %idxprom1`? Should at least have some with additional constant indices?

huntergr-arm replied:
I had a test (`histogram_array_3op_gep`) which exercised this with global arrays, requiring one extra constant index argument.

I've added a second (`histogram_array_4op_gep_nonzero_const_idx`) with a struct as well, with one of the constant terms being nonzero.

Are there other cases you would like to see?

fhahn replied:
Thanks for adding the extra tests.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1048,6 +1052,123 @@ bool LoopVectorizationLegality::canVectorizeInstrs() {
+
+  // Check that the index is calculated by loading from another array. Ignore
+  // any extensions.
+  // FIXME: Support indices from other sources that a linear load from memory?
```
```suggestion
  // FIXME: Support indices from other sources than a linear load from memory?
```

I assume the check `isa<SCEVAddRecExpr>` below guards for this FIXME? Would be good to document this current restriction. Does the step of the SCEV matter? It could also be an add-rec for an other loop, which would pass this check but would be a uniform load?

fhahn replied:
Thanks for adding the new test. 

I might have missed it elsewhere, does the step of the AddRec below matter? Does the loop associated with the add rec matter? e.g. it may be an AddRec in an outer loop, then it won't be incremented on each iteration of the inner loop.

huntergr-arm replied:
I added a check+bailout for a loop-invariant stride below this point. The step doesn't matter beyond that I think. In theory we don't even need it to be an AddRec, but I'm being cautious for now.

huntergr-arm committed:
Addressing review comments:
* Improve test ir names
* Explain current limitation on indices
* Check index address is not loop-invariant (with test)
* Check for null pointer on dependences (with test)
* Remove HistogramInfo from recipe, remove legality header from VPlanRecipes.cpp
* Implement overridden computeCost for VPHistogramRecipe
* Separated VPlan checks from main test file, used regex matching
* Added a 4-op GEP test featuring a struct
* Added 64b histogram test without index extension
* Added comment to target-generic test file explaining it doesn't currently vectorize

SamTebbs33 reviewed (approved):
It looks like you've addressed Florian's remaining comments and I'm happy with the state it's in.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -69,6 +69,7 @@ class LoopVectorizationCostModel;
 class LoopVersioning;
 
 struct VPCostContext;
+struct HistogramInfo;
```
might have missed it, but not used in latest version in the file?

huntergr-arm replied:
done

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1048,6 +1052,133 @@ bool LoopVectorizationLegality::canVectorizeInstrs() {
+    return false;
+
+  if (!isa<SCEVAddRecExpr>(PSE.getSE()->getSCEV(VPtrVal)) ||
+      TheLoop->isLoopInvariant(VPtrVal))
```
Better/clearer to check if the AddRec's loop is TheLoop? I think it may be possible for VPtrVal (the GEP) the be placed inside TheLoop, but still be an AddRec of the outer loop?

huntergr-arm replied:
done

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -1583,6 +1585,50 @@ class VPWidenCallRecipe : public VPSingleDefRecipe {
+/// of intrinsics. The only update operations currently supported are
+/// 'add' and 'sub' where the other term is loop-invariant.
+class VPHistogramRecipe : public VPRecipeBase {
+  unsigned Opcode;
```
nit: document that this is the opcode of the increment?

huntergr-arm replied:
done

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -953,6 +954,92 @@ void VPWidenCallRecipe::print(raw_ostream &O, const Twine &Indent,
+  //        vec-of-pointers.
+  assert(VF.isVector() && "Invalid VF for histogram cost");
+  Value *Address = getOperand(0)->getUnderlyingValue();
+  Value *IncAmt = getOperand(1)->getUnderlyingValue();
```
The operand may be replaced by a recipe without underlying value. To get the type, better to use ` Ctx.Types.inferScalarType`. There's also a ToVector helper to create a vector type

Same for using Address to get the type

huntergr-arm replied:
done

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -953,6 +954,92 @@ void VPWidenCallRecipe::print(raw_ostream &O, const Twine &Indent,
+  // Assume that a non-constant update value (or a constant != 1) requires
+  // a multiply, and add that into the cost.
+  InstructionCost MulCost = TTI::TCC_Free;
+  ConstantInt *RHS = dyn_cast<ConstantInt>(IncAmt);
```
There may be no underlying value for `IncAmt`, better to check if it is a live in (isLiveIn) and use getLiveInIRValue if it is?

huntergr-arm replied:
done

huntergr-arm committed:
Rebase, fix up tests, address comments:
* Check AddRec Loop vs. current
* Remove stray struct token declaration
* Add comment on Opcode field of VPHistogramRecipe
* Use Ctx.Types.inferScalarType when computing cost
* Check for a LiveIn value if there isn't an underlying value for costing
* Tolerate nulls for both underlying and LiveIn, just in case.

huntergr-arm commented:
Rebased, and fixed up a couple of test that had slightly different output.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -6519,8 +6520,33 @@ LoopVectorizationCostModel::getInstructionCost(Instruction *I,
     // We've proven all lanes safe to speculate, fall through.
     [[fallthrough]];
   case Instruction::Add:
+  case Instruction::Sub: {
+    auto Info = Legal->getHistogramInfo(I);
+    if (Info && VF.isVector()) {
+      const HistogramInfo *HGram = Info.value();
+      // Assume that a non-constant update value (or a constant != 1) requires
+      // a multiply, and add that into the cost.
+      InstructionCost MulCost = TTI::TCC_Free;
+      ConstantInt *RHS = dyn_cast<ConstantInt>(I->getOperand(1));
+      if (!RHS || RHS->getZExtValue() != 1)
+        MulCost = TTI.getArithmeticInstrCost(Instruction::Mul, VectorTy);
+
+      // Find the cost of the histogram operation itself.
+      Type *PtrTy = VectorType::get(HGram->Load->getPointerOperandType(), VF);
+      Type *ScalarTy = I->getType();
+      Type *MaskTy = VectorType::get(Type::getInt1Ty(I->getContext()), VF);
+      IntrinsicCostAttributes ICA(Intrinsic::experimental_vector_histogram_add,
+                                  Type::getVoidTy(I->getContext()),
+                                  {PtrTy, ScalarTy, MaskTy});
+
+      // Add the costs together with the add/sub operation.
+      return TTI.getIntrinsicInstrCost(
+                 ICA, TargetTransformInfo::TCK_RecipThroughput) +
+             MulCost + TTI.getArithmeticInstrCost(I->getOpcode(), VectorTy);
+    }
+    [[fallthrough]];
+  }
```
Is this still needed in the latest version with `::computeCost?

```suggestion
```

huntergr-arm replied:
Yes, that's still needed for now -- the two cost models disagree and assert if I remove that code.

fhahn replied:
Thanks for checking, I though the assert wouldn't trigger due to `planContainsAdditionalSimplifications`.

huntergr-arm commented:
I'll try committing this on Friday, unless there are any other comments?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -1047,6 +1048,96 @@ void VPWidenCallRecipe::print(raw_ostream &O, const Twine &Indent,
+  // a multiply, and add that into the cost.
+  Value *RHS = IncAmt->getUnderlyingValue();
+  // The underlying value may be null, check for a live-in if so.
+  if (!RHS && IncAmt->isLiveIn())
```
Is RHS only used to check if it is constant? It can only be constant if it is a live-in, so retrieving getUnderlyingValue first can be skipped?

fhahn commented on llvm/test/Transforms/LoopVectorize/histograms.ll:
```diff
@@ -0,0 +1,46 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 3
+; RUN: opt < %s -passes=loop-vectorize,instcombine -enable-histogram-loop-vectorization -force-vector-width=2 -S | FileCheck %s
+
+target datalayout = "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-s0:64:64-f80:128:128-n8:16:32:64-S128"
```
is this needed?

huntergr-arm committed:
Adapt to new interleaving method

huntergr-arm committed:
Simplify VPHistogramRecipe::computeCost

huntergr-arm committed:
Remove unnecessary text from generic histogram test

huntergr-arm committed:
Use flags instead of metadata

huntergr-arm committed:
Update i8 histogram test now that the backend supports it

huntergr-arm committed:
Allow epilogue vectorization now that the clone() method works on the histogram recipe

huntergr-arm committed:
Formatting

huntergr-arm commented:
Rebased, addressed comments, added support for epilogue vectorization now that the clone method has been implemented. Simplified the tests via flags where appropriate, updated the i8 test since the backend is now able to legalize it.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1051,6 +1055,134 @@ bool LoopVectorizationLegality::canVectorizeInstrs() {
+    return false;
+
+  // Find a single IndirectUnsafe dependency.
+  LAI = &LAIs.getInfo(*TheLoop);
```
Is this needed or can it be dropped? It should be initialized by canVectorizeMemory which is called before this one?

huntergr-arm replied:
done

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -1021,6 +1022,94 @@ void VPWidenCallRecipe::print(raw_ostream &O, const Twine &Indent,
+  if (VPValue *VPMask = getMask())
+    Mask = State.get(VPMask);
+  else
+    Mask = Builder.CreateVectorSplat(VTy->getElementCount(),
+                                     ConstantInt::getTrue(Builder.getInt1Ty()));
```
```suggestion
    Mask = Builder.CreateVectorSplat(VTy->getElementCount(),
                                     Builder.getInt1(1));
```

huntergr-arm replied:
done

fhahn reviewed (approved):
LGTM, thanks!

huntergr-arm committed:
Don't reinitialize LAI

huntergr-arm committed:
Use a shorter way of getting a true value to splat

huntergr-arm commented:
Thanks for the thorough review :)
